amr-wind/boundary_conditions/wall_models/WallFunction.cpp

```diff
@@ -0,0 +1,125 @@
+    : m_wall_func(wall_func)
+{
+    amrex::ParmParse pp("WallFunction");
+    pp.query("wall_shear_stress_type", m_wall_shear_stress_type);
+    m_wall_shear_stress_type = amrex::toLower(m_wall_shear_stress_type);
+
+    if (m_wall_shear_stress_type == "constant") {
+        amrex::Print() << "Shear Stress model: " << m_wall_shear_stress_type
+                       << std::endl;
+    } else {
+        amrex::Abort("Shear Stress wall model input mistake");
+    }
+}
+
+void VelWallFunc::wall_model(
+    Field& velocity, const FieldState rho_state, const amrex::Real utau)
+{
+    BL_PROFILE("amr-wind::VelWallFunc");
+    constexpr int idim = 2;
+    const auto& repo = velocity.repo();
+    const auto& density = repo.get_field("density", rho_state);
+    const auto& viscosity = repo.get_field("velocity_mueff");
+    const int nlevels = repo.num_active_levels();
+
+    amrex::Orientation zlo(amrex::Direction::z, amrex::Orientation::low);
+    amrex::Orientation zhi(amrex::Direction::z, amrex::Orientation::high);
+    if ((velocity.bc_type()[zlo] != BC::wall_model) &&
+        (velocity.bc_type()[zhi] != BC::wall_model)) {
+        return;
+    }
+
+    for (int lev = 0; lev < nlevels; ++lev) {
+        const auto& geom = repo.mesh().Geom(lev);
+        const auto& domain = geom.Domain();
+        amrex::MFItInfo mfi_info{};
+
+        const auto& rho_lev = density(lev);
+        auto& vel_lev = velocity(lev);
+        const auto& eta_lev = viscosity(lev);
+
+        if (amrex::Gpu::notInLaunchRegion()) {
+            mfi_info.SetDynamic(true);
+        }
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        for (amrex::MFIter mfi(vel_lev, mfi_info); mfi.isValid(); ++mfi) {
+            const auto& bx = mfi.validbox();
+            auto varr = vel_lev.array(mfi);
+            auto den = rho_lev.array(mfi);
+            auto eta = eta_lev.array(mfi);
+
+            if (bx.smallEnd(idim) == domain.smallEnd(idim) &&
+                velocity.bc_type()[zlo] == BC::wall_model) {
+                amrex::ParallelFor(
+                    amrex::bdryLo(bx, idim),
+                    [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                        const amrex::Real mu = eta(i, j, k);
+
+                        // Dirichlet BC
+                        varr(i, j, k - 1, 2) = 0.0;
+
+                        // Shear stress BC
+                        varr(i, j, k - 1, 0) = utau * utau / mu * den(i, j, k);
+                        varr(i, j, k - 1, 1) = 0.0;
+                    });
+            }
+
+            if (bx.bigEnd(idim) == domain.bigEnd(idim) &&
+                velocity.bc_type()[zhi] == BC::wall_model) {
+                amrex::ParallelFor(
+                    amrex::bdryHi(bx, idim),
+                    [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                        const amrex::Real mu = eta(i, j, k - 1);
```

Comment:
I realize this is how it's coded now, but I don't think the indices are correct. The bdryHi of the bx gives the highest k inside the domain, as far as I know. This would be the k at the highest cell center. Then mu should be at k, and varr should be set at k + 1. I may be wrong, but could you check this?

Reply:
`amrex::bdryHi(bx, idim)` gives `((0,0,32) (15,15,32) (0,0,1))` so I think this is fine.